### ISSUE TITLE
make completion is slow for huge Makefiles (generated by Automake)

### ISSUE THREAD

**Issue by philipp-classen** [scop/bash-completion#66]
(This is a follow-up on the discussion of #63)

For huge Makefiles, as generated by Automake, bash completions becomes very slow (>10 seconds).

To reproduce, use [huge_makefile_example.zip](https://github.com/scop/bash-completion/files/416801/huge_makefile_example.zip). Completion for make should take several seconds. On my machine, it takes about 4 seconds for the call to "make", and additionally more than 10 seconds for the call to "sed".

The archive also contains the Make database generated by
```
__BASH_MAKE_COMPLETION__=1 make -f Makefile -npq .DEFAULT 2>/dev/null 1>Makefile.db
```

**Comment by SuperSandro2000** [comment]
Is there any progress on this in the meantime? I even encounter a few seconds wait time with very short makefiles which need to run commands for variables. Can command running be disabled to speedup tab completion at the cost of correctnes?

**Comment by akinomyoga** [comment]
I think there are no efforts on this for now.

> with very short makefiles which need to run commands for variables.

In the original report, the time of `sed` (10 secs) seems to be longer than that of `make` (4 secs). In the case that commands are run for variables, I guess the time of `make` should be dominant. What is the result of the following command in the directory?

```bash
$ time make -npq | wc
```

If this explains the most part of the delay you observe, I think there is no way to disable only `$(shell slow-command)` in Makefile unless GNU make supports the feature. However, if we give up completions of all the targets that includes variables or wildcards (such as `$(OBJDIR)/file` or `%.o`) and also give up tracking `include`s, we may just run `sed -n 's/^[[:space:]]*\([^:]*[^[:space:]]\)[[:space:]]*:\([^=].*\)\{0,1\}$/\1/p' Makefile` to extract target names. Or another way might be to use `timeout` command to set a timeout for the call of `make -npq`, and to fall back to the above simple `sed` on the timeout.

**Comment by janev94** [comment]
> I think there are no efforts on this for now.
> 
> > with very short makefiles which need to run commands for variables.
> 
> In the original report, the time of `sed` (10 secs) seems to be longer than that of `make` (4 secs). In the case that commands are run for variables, I guess the time of `make` should be dominant. What is the result of the following command in the directory?
> 
> ```shell
> $ time make -npq | wc
> ```
> 
> If this explains the most part of the delay you observe, I think it's difficult to improve the response because there is no way to disable only `$(shell slow-command)` in Makefile unless GNU make supports the feature. However, if we give up completions of all the targets that includes variables or wildcards (such as `$(OBJDIR)/file` or `%.o`) and also give up tracking `include`s, we may just run `sed -n 's/^[[:space:]]*\([^:]*[^[:space:]]\)[[:space:]]*:\([^=].*\)\{0,1\}$/\1/p' Makefile` to extract target names. Or another way might be to use `timeout` command to set a timeout for the call of `make -npq`, and to fall back to the above simple `sed` on the timeout.

Do you have an example how to instruct make to give up completions of all targets that include variables or wildcards as you suggested, or was that a proposal for how this _might_ be resolved? Asking since I have a similar issue and would love it if I can make the completion faster.

**Comment by akinomyoga** [comment]
> Do you have an example how to instruct make to give up completions of all targets that include variables or wildcards as you suggested,

No. In addition, that's not what I suggested. As I mentioned, *there is no way to do that* unless GNU Make would newly support it. What I instead suggested was to give up using `make` and manually manipulate `Makefile`, `GNUmakefile`, etc. using `sed`, etc.

> or was that a proposal for how this _might_ be resolved?

It was a possibility of how the problem might be fixed in bash-completion, and it's not what I expect the users of bash-completion to set up. The problem with this heuristic approach is that it's incomplete, imprecise, and not robust.
